Use GOOGLE_APPLICATION_CREDENTIALS and GCLOUD_PROJECT vars in RemoteGcsHelper

mziccard opened the pull request:
This fixes #335

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/testing/RemoteGcsHelper.java:
```diff
@@ -145,22 +145,17 @@ public static RemoteGcsHelper create(String projectId, String keyPath)
+   *     variables {@code GCLOUD_PROJECT} and {@code GOOGLE_APPLICATION_CREDENTIALS} are not set or
+   *     if the file pointed by {@code GOOGLE_APPLICATION_CREDENTIALS} does not exist
    */
   public static RemoteGcsHelper create() throws GcsHelperException {
```
Explicit projetId and key verification is fine.

Another option would be to only document that this `create` method is based on the default lookup of [projectId](https://github.com/GoogleCloudPlatform/gcloud-java#specifying-a-project-id) and [credentials] (https://github.com/GoogleCloudPlatform/gcloud-java#authentication) which allows more more flexibility.

Is the current way your preference?

mziccard replied:
I thought about it and wasn't sure how to proceed.
I don't expect much users to run integration tests on top of GCE or AE (is it even possible?).

I don't have a preference here, if you prefer we can use default lookup, I'm happy to change.

aozarov replied:
> I don't expect much users to run integration tests on top of GCE or AE (is it even possible?).

I am pretty sure it is possible but I don't know how common it is.

> I don't have a preference here, if you prefer we can use default lookup, I'm happy to change.

My preference would be to use default lookup.

mziccard replied:
Let me put this on hold until we discuss and possibly close #341

aozarov replied:
I sent #343 to fix this issue

mziccard commented:
I fixed the comment and now RemoteGcsHelper uses default auth credentials.
This should not be merged until we fix default auth credentials NPE.

ajkannan commented:
#344 is now merged in, which fixes the default auth credentials NPE

mziccard commented:
I realized now that using default authentication credentials causes signUrl method to fail (see [here](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java#L587)) and consequently also signUrl IT tests fail.

We can manually pass a `ServiceAccountAuthCredentials` to `signUrl` but that is the same as explicitly creating `ServiceAccountAuthCredentials` from the env variables in `RemoteGcsHelper`.

I suggest that we do not use default lookup in `RemoteGcsHelper` unless you know a way to workaround the signUrl problem

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/testing/RemoteGcsHelper.java:
```diff
@@ -45,8 +45,8 @@
   private static final String BUCKET_NAME_PREFIX = "gcloud-test-bucket-temp-";
-  private static final String PROJECT_ID_ENV_VAR = "GCLOUD_TESTS_PROJECT_ID";
-  private static final String PRIVATE_KEY_ENV_VAR = "GCLOUD_TESTS_KEY";
+  private static final String PROJECT_ID_ENV_VAR = "GCLOUD_PROJECT";
```
These are not used anymore, right?

aozarov commented:
> I realized now that using default authentication credentials causes signUrl method to fail (see here) and consequently also signUrl IT tests fail.

We could also support it when `GoogleCredentials` is instanceof [ServiceAccountCredentials](https://github.com/google/google-auth-library-java/blob/281d485e679545ed06f46e6090fc5d47f4a140e8/oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java). There are various ways we could deal with it.

mziccard commented:
In 82ebc71 I fixed the comment and:
- Added a method to convert `ApplicationDefaultCredentials` to `ServiceAccountAuthCredentials` if `GoogleCredentials` object is an instance of `ServiceAccountCredentials`
- Add type check and conversion to `Storage.signUrl` to work with `ApplicationDefaultCredentials`

aozarov commented:
Great! Looks good to me. Any reason you still have the "don't merge" label?

mziccard commented:
I just forgot to remove :)
